### ISSUE TITLE
Display better security error messages in profiler logs

### ISSUE THREAD

**Issue by tomaszwota** [symfony/symfony#16272]
## In short:

In case the entity class that is referenced in `security.yml` as `provider` is **not** implementing `UserInterface` and `\Serializable` (as it should), there is no trace of the problem in Profiler logs, beyond the generalized "safe" (and intentionally vague) message.
## More details:

Symfony docs on [form login setup](http://symfony.com/doc/current/cookbook/security/form_login_setup.html) provide us by default with a message shown in a template:

``` twig
{{ error.messageKey|trans(error.messageData, 'security') }}
```

which is generally fine, and once you see there's a problem, you can go ahead and echo `error.message` to get a meaningful _"The user provider must return a UserInterface object."_ message.

However, I don't think it should be unreasonable to assume when only a "safe" message is printed, that in `dev` environment we'd also get the more useful and meaningful `error.message` _logged_ in Profiler logs. 

As it stands it's just: _"Authentication request could not be processed due to a system problem."_ in logs, and a newbie, like me, will be confused by the lack of anything more detailed in the Profiler logs.
## Proposed solution:

The idea would be to include "uncensored" `error.message` in Profiler logs in dev environment, when it's appropriate, so that the general assumption of _"if there's a problem, **useful** info should be found in the Profiler logs"_ would hold in this case as it does in so many others.

**Comment by javiereguiluz** [comment]
@tomaszwota thanks for reporting this issue. I've looked into it and I agree with you. However, it's not only a profiler error but a general security error.

### How to reproduce

1) Remove the `implements UserInterface` from your user class and try to login.
2) In the browser you see this:

![security_error](https://cloud.githubusercontent.com/assets/73419/12582722/15b41950-c43f-11e5-91be-d0ded0246b86.png)

3) In the log file you see this:

```
[2016-01-26 15:06:42] request.INFO: Matched route "security_login_form". {"route_parameters":
{"_controller":"AppBundle\\Controller\\SecurityController::loginAction","_locale":"en","_route":"security_l
ogin_form"},"request_uri":"http://symfony-demo.dev/app_dev.php/en/login"} []

[2016-01-26 15:06:42] security.INFO: Populated the TokenStorage with an anonymous Token. [] []
```

### How to fix

The error message displayed to the end-user must be the same ... but Symfony should log better and more detailed security error messages.

---

Anyone wants to work on this improvement? Thanks!

**Comment by WouterJ** [comment]
Let's first very briefly describe what's happening here:

 1. The login page is under a firewall with at least the **`anonymous` and `simple_form` authentication providers enabled**
 1. The login form tries to authenticate the token when logging in. However, the user provider returns an invalid value and the **form authentication provider throws an exception**
 1. On exception, the **authentication manager executes to the next provider** (which is the anonymous provider)
 1. This provider **is able to return a token**
 1. **The token provided by the anonymous provider is populated** and the **Security system is in a success state**
 1. Still, a user does not have the correct rights and the **error of the login form is shown**.

If there was no anonymous authentication provider enabled, you would have the nice log message:

    [2016-02-28 01:13:17] security.INFO: Authentication request failed. {"exception":"[object] (Symfony\\Component\\Security\\Core\\Exception\\AuthenticationServiceException(code: 0): The user provider must return a UserInterface object. at C:\\sf-world\\symfony-28\\vendor\\symfony\\symfony\\src\\Symfony\\Component\\Security\\Core\\Authentication\\Provider\\DaoAuthenticationProvider.php:94, Symfony\\Component\\Security\\Core\\Exception\\AuthenticationServiceException(code: 0): The user provider must return a UserInterface object. at C:\\sf-world\\symfony-28\\vendor\\symfony\\symfony\\src\\Symfony\\Component\\Security\\Core\\Authentication\\Provider\\DaoAuthenticationProvider.php:86)"} []

I'm a bit unsure on how to fix this issue. At the moment, nothing in Security-Core depends on the logger. If we were to log all exceptions thrown by the authentication providers, at least the `AuthenticationProviderManager` has to depend on the logger and log the thrown exceptions.

**Comment by ShinDarth** [comment]
Same here: gettin `Authentication request could not be processed due to a system problem.` but no idea about what is causing the problem.

**Comment by chalasr** [comment]
Actually it seems this has been fixed.
Following the steps to reproduce given by @javiereguiluz on a fresh symfony/demo installation I get the full exception in logs:

![](https://image.prntscr.com/image/B4901nl-RROu_Y-Th8Zt4g.png)

Failing listeners are responsible for [logging their own failure](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Http/Firewall/AbstractAuthenticationListener.php#L196), even though another one succeeds afterwards (e.g. anonymous). 

I reviewed core listeners and they're all good on this point, maybe that's something to mention in the docs for people implementing their own listeners without extending the abstract one (they have to make them log, we can't do it reliably at a lower level, exceptions are caught by listeners, logged and transformed into responses).
